dashboard/test/models/contact_rollups_pardot_memory_test.rb

```diff
@@ -32,52 +32,167 @@ class ContactRollupsPardotMemoryTest < ActiveSupport::TestCase
+    ]
+    processed_contact_records.each {|record| create :contact_rollups_processed, record}
+
+    # Execute SQL query
+    results = ActiveRecord::Base.connection.
+      exec_query(ContactRollupsPardotMemory.query_new_contacts).map do |record|
+      record['email']
+    end
+
+    # Should find only 2 new contacts
+    assert_equal %w(alpha gamma), results
+  end
+
+  test 'create_new_pardot_prospects' do
+    assert_equal 0, ContactRollupsPardotMemory.count
+    contact = create :contact_rollups_processed, data: {'opt_in' => true}
 
-    # Sync multiple contacts
-    processed_contacts.concat(create_list(:contact_rollups_processed, 2))
+    PardotV2.expects(:submit_batch_request).once.returns([])
 
     ContactRollupsPardotMemory.create_new_pardot_prospects
-    assert_equal processed_contacts.length, ContactRollupsPardotMemory.count
-    processed_contacts.each do |contact|
-      refute_nil ContactRollupsPardotMemory.find_by(email: contact.email)
+
+    assert_equal 1, ContactRollupsPardotMemory.count
+    record = ContactRollupsPardotMemory.find_by(email: contact.email)
+    assert_equal({'db_Opt_In' => 'Yes'}, record&.data_synced)
+  end
+
+  test 'query_updated_contacts' do
+    assert_equal 0, ContactRollupsPardotMemory.count
+    assert_equal 0, ContactRollupsProcessed.count
+
+    base_time = Time.now.utc - 2.days
+    pardot_memory_records = [
+      {email: 'alpha', pardot_id: 1, data_synced_at: nil, data_synced: nil},
+      {
+        email: 'beta',
+        pardot_id: 2, pardot_id_updated_at: base_time - 2.days,
+        data_synced_at: base_time - 1.day, data_synced: {db_Opt_In: 'Yes'}
+      },
+      {
+        email: 'gamma',
+        pardot_id: 3, pardot_id_updated_at: base_time + 2.days,
+        data_synced_at: base_time + 1.day, data_synced: {db_Opt_In: 'Yes'}
+      },
+      # dummy records
+      {email: 'delta', pardot_id: nil},
+      {email: 'epsilon', pardot_id: 4},
+    ]
+    pardot_memory_records.each {|record| create :contact_rollups_pardot_memory, record}
+
+    # 3 cases that require updating contacts
+    processed_contact_records = [
+      # Case 1: data_synced_at is null
+      {email: 'alpha', data: {opt_in: false}, data_updated_at: base_time},
+      # Case 2: data_updated_at > data_synced_at
+      {email: 'beta', data: {opt_in: false}, data_updated_at: base_time},
+      # Case 3: pardot_id_updated_at > data_synced_at
+      {email: 'gamma', data: {opt_in: true}, data_updated_at: base_time},
+      # dummy records
+      {email: 'delta'},
+      {email: 'zeta'},
+    ]
+    processed_contact_records.each {|record| create :contact_rollups_processed, record}
+
+    # Execute SQL query
+    results = ActiveRecord::Base.connection.
+      exec_query(ContactRollupsPardotMemory.query_updated_contacts).map do |record|
+      record.slice('email', 'pardot_id_changed')
     end
+
+    expected_results = [
+      {'email' => 'alpha', 'pardot_id_changed' => nil},
```

Comment:
why is pardot_id_changed different for each case here?

Reply:
`pardot_id_changed` is the result of comparing `pardot_id_updated_at` and `data_synced_at`.
* Value of 1 means Pardot ID of an email has changed since the last data sync. (This could happen because we can have multiple Pardot IDs for 1 email.) 
* Value of `nil` means 1 of the 2 operands is null and mysql cannot do null comparison. This may cause some confusion. I'll convert `nil` result to 0.